core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1200 @@
+    (this eq that) || (this.rootNode === that.rootNode)
+
+  final override def equals(that: Any): Boolean = that match {
+    case map: HashMap[_, _] =>
+      (this eq map) || (this.rootNode == map.rootNode)
+    case _ =>
+      false
+  }
+
+  /**
+   * Compute a hash code value for this map.
+   *
+   * This method is similar to [[java.lang.Object#hashCode]] except that it computes a hash code
+   * according to [[cats.Hash]] instances.
+   *
+   * @param hashValue the [[cats.kernel.Hash]] instance to use for hashing values of type `VV`.
+   * @return a hash code value for this map.
+   */
+  final def hash[VV >: V](implicit hashValue: Hash[VV]): Int =
+    StaticMethods.unorderedHash(this.iterator: Iterator[(K, VV)])
+
+  final override def hashCode(): Int = {
+    implicit val valueHash = Hash.fromUniversalHashCode[V]
+    StaticMethods.unorderedHash(this.iterator)
+  }
+
+  /**
+   * Typesafe stringification operator.
+   *
+   * This method is similar to [[java.lang.Object#toString]] except that it stringifies values according
+   * to [[cats.Show]] instances.
+   *
+   * @param showKey the [[cats.Show]] instance to use for showing keys of type `K`.
+   * @param showValue the [[cats.Show]] instance to use for showing values of type `V`.
+   * @return a [[java.lang.String]] representation of this map.
+   */
+  final def show[VV >: V](implicit showKey: Show[K], showValue: Show[VV]): String =
+    iterator.map { case (k, v) => s"${showKey.show(k)} -> ${showValue.show(v)}" }.mkString("HashMap(", ", ", ")")
+
+  final override def toString() =
+    iterator.map { case (k, v) => s"$k -> $v" }.mkString("HashMap(", ", ", ")")
+}
+
+object HashMap extends HashMapInstances {
+  final private[data] def improve(hash: Int): Int =
+    scala.util.hashing.byteswap32(hash)
+
+  /**
+    * Creates a new empty [[cats.data.HashMap]] which uses `hashKey` for hashing.
+    *
+    * @param hashKey the [[cats.kernel.Hash]] instance used for hashing keys.
+    * @return a new empty [[cats.data.HashMap]].
+    */
+  final def empty[K, V](implicit hashKey: Hash[K]): HashMap[K, V] =
+    new HashMap[K, V](Node.empty[K, V])
+
+  /**
+  * Creates a new [[cats.data.HashMap]] which contains all elements of `kvs`.
+  *
+  * @param kvs the key-value pairs to add to the [[cats.data.HashMap]].
+  * @param hashKey the [[cats.kernel.Hash]] instance used for hashing keys.
+  * @return a new [[cats.data.HashMap]] which contains all elements of `kvs`.
+  */
+  final def apply[K, V](kvs: (K, V)*)(implicit hashKey: Hash[K]) =
+    fromSeq(kvs)
+
+  /**
+  * Creates a new [[cats.data.HashMap]] which contains all elements of `seq`.
+  *
+  * @param seq the sequence of elements to add to the [[cats.data.HashMap]].
+  * @param hashKey the [[cats.kernel.Hash]] instance used for hashing values.
+  * @return a new [[cats.data.HashMap]] which contains all elements of `seq`.
+  */
+  final def fromSeq[K, V](seq: Seq[(K, V)])(implicit hashKey: Hash[K]): HashMap[K, V] = {
```

Comment:
why not call `fromFoldable` here? There is a `Foldable[Seq]` no?

Reply:
This is a good observation actually - this currently accepts a (potentially mutable) `scala.collection.Seq` on 2.12, which has no `Foldable` instance.

Do we keep the implementation of `fromSeq` here or restrict it to `collection.immutable.Seq`?